include/swift/AST/Expr.h

```diff
@@ -4593,10 +4594,10 @@ class DotSyntaxBaseIgnoredExpr : public Expr {
 class ExplicitCastExpr : public Expr {
   Expr *SubExpr;
   SourceLoc AsLoc;
-  TypeExpr *const CastTy;
+  TypeLoc CastTy;
 
 protected:
-  ExplicitCastExpr(ExprKind kind, Expr *sub, SourceLoc AsLoc, TypeExpr *castTy)
+  ExplicitCastExpr(ExprKind kind, Expr *sub, SourceLoc AsLoc, TypeRepr *castTy)
```

Comment:
I was a big proponent of using `TypeRepr` here and other expressions when @CodaFi worked on switching off of `TypeLoc` but, as we have discovered, the problem is that there could be a difference in type between cast expression itself and its `CastTy` and both types are needed by later compiler stages so it has been decided to use `TypeExpr` here to store a type but hide it so it can't be referenced directly.